### PR TITLE
Modified the CMakeLists.txt such that it generates mlpack-config.cmake

### PR DESCRIPTION
**Pull request by amirmasoudabdol**
I recently ran into some problems integrating mlpack to my projects. I don't know what has changed in past few months, but my old CMake setup couldn't find mlpack anymore. So, I did some digging and decided to give the `CMakeLists.txt` some love.

I realized that the problem is the lack of `findMLPACK.cmake`. So, I started investigating and added some lines to the script to generate to generate `mlpack-config.cmake`. This is very useful since it helps CMake to find mlpack much easier; therefore, makes it easier to integrate it to other projects.

I've tested this on macOS Catalina 10.15.4 and I can integrate mlpack into my projects by just adding `find_package(mlpack)` and including the header files using `include_directories(${MLPACK_INCLUDE_DIRS})`. There were some complication around the `${MLPACK_INCLUDE_DIRS}` but I think I have it right now. 

I'm wondering how everyone thinks about these changes, and whether all tests pass or not. If everyone thinks this is a good idea, we can continue cleaning up the `CMakeLists.txt` and improve the build process further. Also, please fix any stupid mistakes I made. I mainly tried to read other projects, e.g., `fmt`, and study CMake manual as much as I can.

### CODE REVIEW

**Comment by amirmasoudabdol** [comment]
I see that I missed a `if (UNIX)` for `include(GNUInstallDirs)`. So, I added it, and checking if Windows build can be successful now as well.

**Comment by birm** [comment]
We haven't touched it in a while, but there is a [FindMLPACK.cmake](https://github.com/mlpack/models/blob/master/CMake/FindMLPACK.cmake) file in the models repository. That said, this idea seems like a good addition too.

**Comment by amirmasoudabdol** [comment]
I see that some of the jobs are "canceled", should I resubmit the RP to trigger them again? It seems Azure was busy or something...

File: CMakeLists.txt [hunk 1]
```diff
@@ -727,4 +727,82 @@ if (PKG_CONFIG_FOUND)
...
+  install(EXPORT ${targets_export_name}
+          DESTINATION ${MLPACK_CMAKE_DIR} NAMESPACE mlpack::)
+
+  # Install the library and headers.
+  install(TARGETS ${INSTALL_TARGETS} EXPORT ${targets_export_name}
+          LIBRARY DESTINATION ${MLPACK_LIBS_DIR}
+          ARCHIVE DESTINATION ${MLPACK_LIBS_DIR}
+          RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR})
+  
+endif ()
```
Comment (kartikdutt18) [inline CMakeLists.txt:808]: Hey @amirmasoudabdol, you could add an empty line here that would fix the style issue as well as re-trigger the build. Thanks a lot.

File: CMake/mlpack-config.cmake.in [hunk 2]
```diff
@@ -0,0 +1,7 @@
+@PACKAGE_INIT@
+
+include(${CMAKE_CURRENT_LIST_DIR}/@targets_export_name@.cmake)
+
+set_and_check(MLPACK_INCLUDE_DIRS "@PACKAGE_MLPACK_INCS_DIR@")
+
+set(MLPACK_LIBRARIES mlpack::mlpack)
```
Comment (kartikdutt18) [inline CMake/mlpack-config.cmake.in:7]: Hey, Could you add an empty line here as well. Sorry, tiny style fix but that should re-trigger the build. :)

File: CMakeLists.txt (hunk 1 above)
Comment (amirmasoudabdol) [reply CMakeLists.txt:808]: Sure. Please let me know if you see anything else being odd or out of place :-)

**Comment by amirmasoudabdol** [comment]
@birm I noticed that one but it's not placed correctly and CMake cannot use it in other projects to find mlpack correctly. To elaborate for everyone what I exactly did, I utilized the [CMakePackageConfigHelpers](http://cmake.org/cmake/help/v3.17/module/CMakePackageConfigHelpers.html#command:configure_package_config_file) module to generate some CMake for CMake. It generates a few files, e.g., `mlpack-config-target.cmake` and `mlpack-config-version.cmake`, and later when `find_package(mlpack)` is called, it looks for these and setup the variables based on the system pref, ....

**Comment by birm** [comment]
I think this may be a good idea to add a CI test for; ensure some simple example can find mlpack via cmake properly after this has been run. If you don't want to do that, could you/someone test on windows?

**Comment by amirmasoudabdol** [comment]
I agree, I was even thinking to add some documentation on how to integrate mlpack to a CMake project. I think this will be useful too.

I can try doing that later too, for now, I see that some of the Jenkins' tests have not passed and they are concerning the installation process. I'm gonna investigate those first. I need to setup a few VM to test this on Linux systems it seems.

**Comment by amirmasoudabdol** [comment]
So, I investigated this a bit further and tried to test it on a clean Ubuntu and I ran into an interesting problem. If I install the mlpack manually with the modified `CMakeLists.txt`, then CMake doesn't have any problem finding mlpack but it cannot find `ensmallen.hpp`. This is because I didn't install it prior to installing the mlpack, and I was thinking that the current CMake takes care of it but apparently it didn't.

I didn't ran into this problem on my Mac because I already had Ensmallen installed in the `/usr/local` and CMake could find it properly. Another reason for this is that most package managers are installing dependencies first and then get to the library itself. I think this is the case for Homebrew as well, so the package manager hides the current CMake's issue.

Unless I'm missing something, this tells me that the current build system doesn't handle things precisely and there are hardcoded parts that might cause problems in different systems. Of course, this was not apparent because the hacky setup already works, but if we want to change things, I guess we need to dig a bit deeper and identify those. 

Regarding the Ensmallen, as far as I can see, the library has a proper `CMakeLists.txt`, and it takes care of its own install. I don't know about the STB_IMAGE yet. My suggestion is to first make sure that we are installing the dependencies correctly, and see where we get from there.

PS., That being said, I don't understand why those Jenkin's build are missing, I see that it's a CMake error but I don't think I have enough information to reproduce them. If someone could help me a bit I can probably look a bit further.

**Comment by amirmasoudabdol** [comment]
I checked the STB_IMAGE and if I'm correct you can embed it as a single header file as well. That only leaves the Ensmallen as a dependency that needs to be downloaded and installed. I see that @birm fixed a bug in Ensmallen `CMakeLists.txt`, and I tested the fix and can confirm that CMake can find it properly. I also tested mlpack installation both on macOS and Ubuntu with the ensmallen HEAD and modified `CMakeLists.txt` of this PR, and everything seems to be working fine. 

I'm wondering about the `find_package(StbImage)` though. I didn't find any CMake scripts in the project, and not even in debian `apt` directory, so, I'm wondering if this could ever find `StbImage` at all? I don't see it being configured via `pkg-config` either. That being said, I didn't ran into any problem during the installation.

**Comment by rcurtin** [comment]
Hey @amirmasoudabdol, thank you for looking into this!  I totally agree that we could improve the CMake configuration and it would be really nice to be able to provide a properly-configured `mlpack-config.cmake` file for sure. :+1:

> So, I investigated this a bit further and tried to test it on a clean Ubuntu and I ran into an interesting problem. If I install the mlpack manually with the modified CMakeLists.txt, then CMake doesn't have any problem finding mlpack but it cannot find ensmallen.hpp. This is because I didn't install it prior to installing the mlpack, and I was thinking that the current CMake takes care of it but apparently it didn't.

Hmm, some time back the CMake configuration was updated so that it installs ensmallen if ensmallen was automatically downloaded: https://github.com/mlpack/mlpack/blob/master/CMakeLists.txt#L390

Same with STB: https://github.com/mlpack/mlpack/blob/master/CMakeLists.txt#L328

If you run on the master branch, does the problem still occur?  Maybe there is an additional bug there or something...

> PS., That being said, I don't understand why those Jenkin's build are missing, I see that it's a CMake error but I don't think I have enough information to reproduce them. If someone could help me a bit I can probably look a bit further.

My guess is that this has to do with the version of CMake.  On those machines:

```
$ cmake --version
cmake version 3.10.2

CMake suite maintained and supported by Kitware (kitware.com/cmake).
```

That's not ancient, but maybe it's not the same version as you were testing with?

> I'm wondering about the find_package(StbImage) though. I didn't find any CMake scripts in the project, and not even in debian apt directory, so, I'm wondering if this could ever find StbImage at all? I don't see it being configured via pkg-config either. That being said, I didn't ran into any problem during the installation.

The script to find that should be found in `CMake/FindStbImage.cmake`; we have a custom CMake module path that we append so that we can make sure some package search scripts are always available.

Anyway, I'll leave some comments on the PR too, but I am pretty new to the CMake functionality you're using so I think I will have to do a little learning too as I review it. :)

**Comment by amirmasoudabdol** [comment]
Hey @rcurtin,

I didn't notice the `FindStbImage.cmake` but I'll look into it. However, what do you think of just including it as a standalone header if it's possible, then you don't have to deal with it? I'm just wondering, I'm sure there are reasons to keep it like that.

Regarding the failed builds, you are right, it's because of the CMake. As far as I understood it, CMakes prior to 3.13 cannot install targets from different directories. This is being added in 3.13. I assume we want to keep the CMake requirement low, and that means we need to refactor parts of the CMake script. Check these for reference, [here](https://cmake.org/Bug/view.php?id=14444), [here](https://stackoverflow.com/questions/34443128/cmake-install-targets-in-subdirectories), and [here](https://stackoverflow.com/questions/3547331/installtargets-and-add-subdirectory?noredirect=1&lq=1).

It seems we have a few options, one would be to move the mlpack installation (basically the added part in this PR) into `src/mlpack/CMakeListts.txt` but that means we might need to change a few more things. I have already tested this (with CMake 3.10.2), and I can see that older CMake can handle it and the installation is correct. I'm going to commit the changes and see how Jenkins would do. I don't expect this to cause a lot of trouble since the CMake Helper script is pretty independent, and it seems CMake does a good job here. The only extra change that I needed to do is to extract mlpack's version, so I just took it out of the `if (BUILDING_PYTHON_BINDINGS)`. I think it's good to have it anyway.

P.S. As far as I understood this is also partially because of the way mlpack project is setup. Basically, if mlpack was not in a subdirectory, we would not have this issue with older CMakes. I believe this is the way Ensmallen and Armadillo are.

**Comment by amirmasoudabdol** [comment]
So, as far as I can tell, the patch has fixed the issue with Jenkin's CMake and build and tests are successful on them too. I see that some of the old tests are failed but if I'm right they just ran out of time. I see the CMake build is successful, but somehow got stuck. I think I should just make a dummy push to re-trigger the build, right?

**Comment by kartikdutt18** [comment]
>  I see that some of the old tests are failed but if I'm right they just ran out of time. I see the CMake build is successful, but somehow got stuck.

Right, A dummy commit will re-trigger all builds (though nearly all builds are passing, I don't see any reason why linux plain will fail).

File: CMakeLists.txt [hunk 3]
```diff
@@ -727,4 +727,6 @@ if (PKG_CONFIG_FOUND)
 
   install(FILES "${CMAKE_CURRENT_BINARY_DIR}/lib/pkgconfig/mlpack.pc"
       DESTINATION "${CMAKE_INSTALL_LIBDIR}/pkgconfig/")
+
+
```
Comment (rcurtin) [inline CMakeLists.txt:731]: Accidental newlines? :)

File: src/mlpack/CMakeLists.txt [hunk 4]
```diff
@@ -158,3 +159,80 @@ endif ()
...
+# Sets a cache variable with a docstring joined from multiple arguments:
+#   set(<variable> <value>... CACHE <type> <docstring>...)
+# This allows splitting a long docstring for readability.
+function(set_verbose)
+cmake_parse_arguments(SET_VERBOSE "" "" "CACHE" ${ARGN})
+list(GET SET_VERBOSE_CACHE 0 type)
+list(REMOVE_AT SET_VERBOSE_CACHE 0)
+join(doc ${SET_VERBOSE_CACHE})
+set(${SET_VERBOSE_UNPARSED_ARGUMENTS} CACHE ${type} ${doc})
+endfunction()
```
Comment (rcurtin) [inline src/mlpack/CMakeLists.txt:183]: Maybe we should put these into a separate file in `CMake/` and then include them?

File: src/mlpack/CMakeLists.txt [hunk 5]
```diff
@@ -158,3 +159,80 @@ endif ()
...
+
+include(CMakePackageConfigHelpers)
+
+set_verbose(MLPACK_CMAKE_DIR ${CMAKE_INSTALL_LIBDIR}/cmake/mlpack CACHE STRING
+          "Installation directory for cmake files, relative to "
+          "${CMAKE_INSTALL_PREFIX}.")
+set(version_config ${PROJECT_BINARY_DIR}/mlpack-config-version.cmake)
+set(project_config ${PROJECT_BINARY_DIR}/mlpack-config.cmake)
+set(targets_export_name mlpack-targets)
```
Comment (rcurtin) [inline src/mlpack/CMakeLists.txt:192]: Just for consistency, do you mind making the CMake variables all-caps?  All of the other CMake variables (or, well, most of them) are like that. :+1:

File: src/mlpack/CMakeLists.txt [hunk 6]
```diff
@@ -158,3 +159,80 @@ endif ()
...
+set(result )
+foreach (arg ${ARGN})
+  set(result "${result}${arg}")
+endforeach ()
+set(${result_var} "${result}" PARENT_SCOPE)
+endfunction()
+
+# Sets a cache variable with a docstring joined from multiple arguments:
+#   set(<variable> <value>... CACHE <type> <docstring>...)
+# This allows splitting a long docstring for readability.
```
Comment (rcurtin) [inline src/mlpack/CMakeLists.txt:176]: This is awesome!  Thank you.  Much needed. :+1:

File: src/mlpack/CMakeLists.txt [hunk 7]
```diff
@@ -158,3 +159,80 @@ endif ()
...
+if (UNIX)
+GNUInstallDirs_get_absolute_install_dir(MLPACK_LIBS_DIR CMAKE_INSTALL_LIBDIR)
+GNUInstallDirs_get_absolute_install_dir(MLPACK_INCS_DIR CMAKE_INSTALL_INCLUDEDIR)
+endif()
+
+# Generate the version, config and target files into the build directory.
+write_basic_package_version_file(
+${version_config}
+VERSION ${PACKAGE_VERSION}
+COMPATIBILITY SameMajorVersion)
```
Comment (rcurtin) [inline src/mlpack/CMakeLists.txt:215]:
```suggestion
write_basic_package_version_file(
    ${version_config}
    VERSION ${PACKAGE_VERSION}
    COMPATIBILITY SameMajorVersion)
```
Just some formatting. :)  We don't have the style check job for CMake but I can point it out manually, at risk of being pedantic.  Sorry about that. :)

File: src/mlpack/CMakeLists.txt [hunk 8]
```diff
@@ -158,3 +159,80 @@ endif ()
...
+
+configure_package_config_file(
+${PROJECT_SOURCE_DIR}/CMake/mlpack-config.cmake.in
+${project_config}
+INSTALL_DESTINATION ${MLPACK_CMAKE_DIR}
+PATH_VARS MLPACK_INCS_DIR)
```
Comment (rcurtin) [inline src/mlpack/CMakeLists.txt:221]:
```suggestion
configure_package_config_file(
    ${PROJECT_SOURCE_DIR}/CMake/mlpack-config.cmake.in
    ${project_config}
    INSTALL_DESTINATION ${MLPACK_CMAKE_DIR}
    PATH_VARS MLPACK_INCS_DIR)
```
:)

File: src/mlpack/CMakeLists.txt (hunk 4 above)
Comment (amirmasoudabdol) [reply src/mlpack/CMakeLists.txt:183]: Sure, there are not essential functions anyway. I check maybe I just get rid of them.

File: src/mlpack/CMakeLists.txt (hunk 5 above)
Comment (amirmasoudabdol) [reply src/mlpack/CMakeLists.txt:192]: Yes, I'll do that :-)

**Comment by amirmasoudabdol** [comment]
Thanks for your comments @rcurtin. I think I addressed them all. I like that this is progressing :-)

> You mentioned that the install commands now need to be in src/mlpack/CMakeLists.txt. I am not sure on this (maybe the answer is no) but could we put them in the root CMakeLists.txt, but after the src/mlpack/CMakeLists.txt file is included?

As far as I can tell we need to have it there. There are workaround but none looked very appealing to me, they were somewhat hacky, and that means we'll take things out of CMake's hand, and I think that'll go wrong at some point in the future again. But, if someone has a nice way of addressing it, why not. While it's a bit weird that the installer should be in the subdirectories, it somehow makes sense tough. Basically, where we build `mlpack` we prepare its installation, ... as well. I think Even PkgConfig stuff can go into `src/mlpack/CMakeLists.txt`, but again I also understand why this is unsettling. You want the highest level CMake handles the installation, etc.

> I am not sure, maybe it could be restructured, but the main difference between ensmallen and armadillo's file structure is that with mlpack, the expectation is that you'll include <mlpack/core.hpp>, not something in the root of the include directory (like #include <armadillo> or #include <ensmallen.hpp>). So I don't know how we could move it around. If you have an idea, I don't think anyone is hugely tied to the current structure. :)

That's exactly what older CMake cannot handle for somer reason, they cannot install libraries in subdirectories! 🤷‍♂️ I don't think we need to restructure the entire `mlpack`, if that's what you meant! 😅

File: CMake/CMakeExtras.cmake [hunk 9]
```diff
@@ -0,0 +1,21 @@
+# Set of helper functions for our CMake
+
+# Joins arguments and places the results in ${result_var}.
+function(join result_var)
+  set(result )
```
Comment (rcurtin) [inline CMake/CMakeExtras.cmake:5]:
```suggestion
  set(result)
```
This comment is basically the pinnacle of pedantry. :)

File: src/mlpack/CMakeLists.txt [hunk 10]
```diff
@@ -158,3 +159,62 @@ endif ()
 # If we are building Markdown documentation, we have to run some setup after we
 # recurse into methods/.  If not, this function is empty.
 post_markdown_setup()
+
+# Installing CMake Helper Files
```
Comment (rcurtin) [inline src/mlpack/CMakeLists.txt:163]: Maybe it's worth changing this comment to point out that this block of code is focused on generating `mlpack-config.cmake` and related files?

**Comment by rcurtin** [comment]
Also, I'm not worried about the build failure; it seems unrelated.

File: src/mlpack/CMakeLists.txt (hunk 10 above)
Comment (birm) [reply src/mlpack/CMakeLists.txt:163]:
```suggestion
# Installing CMake Helper Files for mlpack-config.cmake
```

Comment (birm) [reply src/mlpack/CMakeLists.txt:163]: These are both minor changes, I'll just add them then approve/merge.

**Pull request merged**